### ISSUE TITLE
Compatibility with Carthage broken by OCMock inclusion via CocoaPods

### ISSUE THREAD

**Issue by dmazzoni** [Brightify/Cuckoo#309]
It seems that version 1.2.0 can no longer be integrated with Carthage due to the transitive dependency on OCMock, for the following reasons:

-  Cuckoo does not expose the transitive dependency in a ```Cartfile```, so Carthage doesn't know that OCMock is required
- Even if the ```Cartfile``` was added, building the Cuckoo+OCMock schemes would fail because the OCMock framework path specified in the project file is different from the path where Carthage places the built dependencies

I'm not sure how to fix the second issue while keeping both CocoaPods and Carthage compatibility at the same time.

**Comment by jamezilla** [comment]
Just wondering if OCMock support wouldn't make more sense as a separate project with Cuckoo as a dependency? That way Cuckoo can stay slim for pure swift projects, and this Carthage issue can be sidestepped by Cuckoo adopters for the time-being?

**Comment by MatyasKriz** [comment]
Hey, @jamezilla, that's a good idea, thanks!
One thing that we need to figure out is how to join Cuckoo and OCMock for testing, so that developers don't have to run tests twice - once for Cuckoo and once for OCMock. Although one might argue that they are different enough to not affect one another.
Also, given how simple Carthage is, it doesn't really give us a say about which source files to include. I'm wondering if it will automatically include a submodule if we make OCMock into one. What do you think?

**Comment by Sajjon** [comment]
Confirming that version `1.2.0` does not build with Carthage, on macOS Catalina, Xcode 11.1, error log (`/var/folders/qp/6xfqjd090yb93vh59t5ttkb80000gn/T/carthage-xcodebuild.ChWtxx.log`):

```
Signing Identity:     "-"

    /usr/bin/codesign --force --sign - --timestamp=none ~/Library/Caches/org.carthage.CarthageKit/DerivedData/11.1_11A1027/Cuckoo/1.2.0/Build/Products/Release-iphonesimulator/Cuckoo.framework

** BUILD SUCCEEDED ** [0.189 sec]

/usr/bin/xcrun xcodebuild -project SwiftProjectDirectory/Carthage/Checkouts/Cuckoo/Cuckoo.xcodeproj -scheme Cuckoo+OCMock-iOS -configuration Release -derivedDataPath ~/Library/Caches/org.carthage.CarthageKit/DerivedData/11.1_11A1027/Cuckoo/1.2.0 -sdk iphoneos ONLY_ACTIVE_ARCH=NO CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES archive -archivePath /var/folders/qp/6xfqjd090yb93vh59t5ttkb80000gn/T/Cuckoo SKIP_INSTALL=YES GCC_INSTRUMENT_PROGRAM_FLOW_ARCS=NO CLANG_ENABLE_CODE_COVERAGE=NO STRIP_INSTALLED_PRODUCT=NO (launched in SwiftProjectDirectory/Carthage/Checkouts/Cuckoo)User defaults from command line:
    IDEArchivePathOverride = /var/folders/qp/6xfqjd090yb93vh59t5ttkb80000gn/T/Cuckoo
    IDEDerivedDataPathOverride = ~/Library/Caches/org.carthage.CarthageKit/DerivedData/11.1_11A1027/Cuckoo/1.2.0

Build settings from command line:
    CARTHAGE = YES
    CLANG_ENABLE_CODE_COVERAGE = NO
    CODE_SIGN_IDENTITY = 
    CODE_SIGNING_REQUIRED = NO
    GCC_INSTRUMENT_PROGRAM_FLOW_ARCS = NO
    ONLY_ACTIVE_ARCH = NO
    SDKROOT = iphoneos13.1
    SKIP_INSTALL = YES
    STRIP_INSTALLED_PRODUCT = NO

note: Using new build system
note: Planning build
note: Constructing build description
error: SwiftProjectDirectory/Carthage/Checkouts/Cuckoo/Pods/Target Support Files/Pods-Cuckoo+OCMock-iOS/Pods-Cuckoo+OCMock-iOS.release.xcconfig: unable to open file (in target "Cuckoo+OCMock-iOS" in project "Cuckoo") (in target 'Cuckoo+OCMock-iOS' from project 'Cuckoo')
error: SwiftProjectDirectory/Carthage/Checkouts/Cuckoo/Pods/Target Support Files/Pods-Cuckoo+OCMock-iOS/Pods-Cuckoo+OCMock-iOS.release.xcconfig: unable to open file (in target "Cuckoo+OCMock-iOS" in project "Cuckoo") (in target 'Cuckoo+OCMock-iOS' from project 'Cuckoo')
error: SwiftProjectDirectory/Carthage/Checkouts/Cuckoo/Pods/Target Support Files/Pods-Cuckoo+OCMock-iOS/Pods-Cuckoo+OCMock-iOS.release.xcconfig: unable to open file (in target "Cuckoo+OCMock-iOS" in project "Cuckoo") (in target 'Cuckoo+OCMock-iOS' from project 'Cuckoo')

** ARCHIVE FAILED **
```

**Comment by MatyasKriz** [comment]
Hey, everyone. Yesterday I was trying to see if Carthage grew up already and allows a simple whitelist/blacklist for specifying which schemes are to be used and which are not necessary for users to build.

I came up empty-handed. No matter what, Carthage just builds ALL shared schemes. For easy development, the OCMock integration is shared as well, leading to the source of this issue.

The workaround:
```
carthage update --no-build && pushd Carthage/Checkouts/Cuckoo/Cuckoo.xcodeproj/xcshareddata/xcschemes && find . ! -name "Cuckoo-iOS.xcscheme" ! -name "Cuckoo-macOS.xcscheme" -delete && popd && carthage build
```

A lot of issues and pull requests have been created in Carthage to solve this very issue, but every time it just gets rejected without any further advice as to what would help get this feature added.

We have started preparing for a new way to manage the Xcode project that will solve this once and for all. I can't tell when this will be ready for release; until that, if you need higher version of Cuckoo to get a feature that would help or just get rid of errors, I recommend using Cocoapods. If that's not an option, calling the command above instead of the usual `carthage update` is the alternative.

**Comment by sanoj00b** [comment]
@MatyasKriz 
Even I am facing the this issue, With Xcode 11.1.2 It not building.

*** Building scheme "Cuckoo+OCMock-iOS" in Cuckoo.xcodeproj
```
Build Failed
    Task failed with exit code 65:
    /usr/bin/xcrun xcodebuild -project /Users/****/Documents/Workspace/***/Carthage/Checkouts/Cuckoo/Cuckoo.xcodeproj -scheme Cuckoo+OCMock-iOS -configuration Release -derivedDataPath /Users/****/Library/Caches/org.carthage.CarthageKit/DerivedData/11.2.1_11B500/Cuckoo/1.3.0 -sdk iphoneos ONLY_ACTIVE_ARCH=NO CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES archive -archivePath /var/folders/82/_m5xwhxx45wyxlhtw40ch509_hqq/T/Cuckoo SKIP_INSTALL=YES GCC_INSTRUMENT_PROGRAM_FLOW_ARCS=NO CLANG_ENABLE_CODE_COVERAGE=NO STRIP_INSTALLED_PRODUCT=NO (launched in /Users/****/Documents/Workspace/crystal/Carthage/Checkouts/Cuckoo)

This usually indicates that project itself failed to compile. Please check the xcodebuild log for more details: /var/folders/82/_m5xwhxx4795wxlhtw40ch509_hqq/T/carthage-xcodebuild.StIE4q.log
```

In the log file it is failing while extracting, please let me know the fix or workaround.
```
** BUILD SUCCEEDED **

/usr/bin/xcrun xcodebuild -project /Users/****/Documents/Workspace/****/Carthage/Checkouts/Cuckoo/Cuckoo.xcodeproj -scheme Cuckoo+OCMock-iOS -configuration Release -derivedDataPath /Users/****/Library/Caches/org.carthage.CarthageKit/DerivedData/11.2.1_11B500/Cuckoo/1.3.0 -sdk iphoneos ONLY_ACTIVE_ARCH=NO CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES archive -archivePath /var/folders/82/_m5xwhxx4795wyxlhtw40ch509_hqq/T/Cuckoo SKIP_INSTALL=YES GCC_INSTRUMENT_PROGRAM_FLOW_ARCS=NO CLANG_ENABLE_CODE_COVERAGE=NO STRIP_INSTALLED_PRODUCT=NO (launched in /Users/****/Documents/Workspace/****/Carthage/Checkouts/Cuckoo)User defaults from command line:
    IDEArchivePathOverride = /var/folders/82/_m5xwhxx4795wyxlhtw40ch509_hqq/T/Cuckoo
    IDEDerivedDataPathOverride = /Users/****/Library/Caches/org.carthage.CarthageKit/DerivedData/11.2.1_11B500/Cuckoo/1.3.0

Build settings from command line:
    CARTHAGE = YES
    CLANG_ENABLE_CODE_COVERAGE = NO
    CODE_SIGN_IDENTITY = 
    CODE_SIGNING_REQUIRED = NO
    GCC_INSTRUMENT_PROGRAM_FLOW_ARCS = NO
    ONLY_ACTIVE_ARCH = NO
    SDKROOT = iphoneos13.2
    SKIP_INSTALL = YES
    STRIP_INSTALLED_PRODUCT = NO

note: Using new build system
note: Planning build
note: Constructing build description
error: /Users/****/Documents/Workspace/****/Carthage/Checkouts/Cuckoo/Pods/Target Support Files/Pods-Cuckoo+OCMock-iOS/Pods-Cuckoo+OCMock-iOS.release.xcconfig: unable to open file (in target "Cuckoo+OCMock-iOS" in project "Cuckoo") (in target 'Cuckoo+OCMock-iOS' from project 'Cuckoo')
error: /Users/****/Documents/Workspace/****/Carthage/Checkouts/Cuckoo/Pods/Target Support Files/Pods-Cuckoo+OCMock-iOS/Pods-Cuckoo+OCMock-iOS.release.xcconfig: unable to open file (in target "Cuckoo+OCMock-iOS" in project "Cuckoo") (in target 'Cuckoo+OCMock-iOS' from project 'Cuckoo')
error: /Users/****/Documents/Workspace/****/Carthage/Checkouts/Cuckoo/Pods/Target Support Files/Pods-Cuckoo+OCMock-iOS/Pods-Cuckoo+OCMock-iOS.release.xcconfig: unable to open file (in target "Cuckoo+OCMock-iOS" in project "Cuckoo") (in target 'Cuckoo+OCMock-iOS' from project 'Cuckoo')

** ARCHIVE FAILED **
```

**Comment by MatyasKriz** [comment]
@sanoj00b the workaround is literally the comment above yours.

**Comment by MatyasKriz** [comment]
Hello, everyone!

I'm back with great news. We've experimentally migrated Cuckoo to use [tuist](https://github.com/tuist/tuist) which allows for not sharing schemes which steps around the whole "Carthage can't whitelist" problem.

Before we merge it and release it, I'd like to ask all of you to try the branch `feature/tuist` to make sure that it's ready to be used in the field in different environments.

When using `feature/tuist`, you needn't use the one-liner I offered above, you may simply call `carthage update`.

Thanks for your patience and thanks for testing the branch out! I'm looking forward to merging this to `master`.

**Comment by JZDesign** [comment]
On `feature/tuist` Carthage `update` and `bootstrap` worked for me! Thanks.

**Comment by MatyasKriz** [comment]
Good to hear, @JZDesign and thanks for the feedback! I'll wait a bit longer to see if anyone encountered any issues using this branch and if not, we'll release this along with some other smaller fixes.

**Comment by kainosk** [comment]
Thank you for `feature/tuist` !!!

I had used CocoaPods because Carthage support was broken. Now, we can use this branch perfectly. 

Thank you ❤️

**Comment by MatyasKriz** [comment]
I'm happy to hear that, @kainosk. Thank you for letting me know. 🙂 I'm preparing a release where this will be included, so it will then not be necessary to point to the specific branch. I'll let everyone here know when it's out.

**Comment by MatyasKriz** [comment]
Released in `1.4.0`! If there are any issues with the latest release, let me know here if it's related to OCMock or in a new issue if it's something else. Thanks!
